core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,269 @@
+    /// - Its own healthcheck is ignored (the task is expected to handle its own checks itself).
+    /// - There may be only one healthcheck task per node.
+    ///
+    /// If the healthcheck task is already set, the provided one will override it.
+    pub fn with_healthcheck<
+        F: FnOnce(
+                &NodeContext<'_>,
+                Vec<Box<dyn CheckHealth>>,
+            ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>
+            + 'static,
+    >(
+        &mut self,
+        healthcheck_task_constructor: F,
+    ) -> &mut Self {
+        self.healthcheck_task_constructor = Some(Box::new(healthcheck_task_constructor));
+        self
+    }
+
+    /// Runs the system.
+    pub fn run(mut self) -> anyhow::Result<()> {
+        // Initialize tasks.
+        let task_constructors = std::mem::take(&mut self.task_constructors);
+
+        let mut tasks = Vec::new();
+        let mut healthchecks = Vec::new();
+
+        let mut errors: Vec<(String, TaskInitError)> = Vec::new();
+
+        for (name, task_constructor) in task_constructors {
+            let mut task = match task_constructor(&NodeContext::new(&self)) {
+                Ok(task) => task,
+                Err(err) => {
+                    // We don't want to bail on the first error, since it'll provide worse DevEx:
+                    // People likely want to fix as much problems as they can in one go, rather than have
+                    // to fix them one by one.
+                    errors.push((name, err));
+                    continue;
+                }
+            };
+            let healthcheck = task.healthcheck();
+            let after_node_shutdown = task.after_node_shutdown();
+            let task_future = Box::pin(task.run(self.stop_receiver.clone()));
+            let task_repr = TaskRepr {
+                name,
+                task: Some(task_future),
+                after_node_shutdown,
+            };
+            tasks.push(task_repr);
+            if let Some(healthcheck) = healthcheck {
+                healthchecks.push(healthcheck);
+            }
+        }
+
+        // Report all the errors we've met during the init.
+        if !errors.is_empty() {
+            for (task, error) in errors {
+                tracing::error!("Task {task} can't be initialized: {error}");
+            }
+            anyhow::bail!("One or more task weren't able to start");
+        }
+
+        if tasks.is_empty() {
+            anyhow::bail!("No tasks to run");
+        }
+
+        // Initialize the healthcheck task, if any.
+        if let Some(healthcheck_constructor) = self.healthcheck_task_constructor.take() {
+            let healthcheck_task = healthcheck_constructor(&NodeContext::new(&self), healthchecks)
+                .map_err(|err| {
+                    tracing::error!("Healthcheck task can't be initialized: {err}");
+                    anyhow::format_err!("Healtcheck task can't be initialized: {err}")
```

Comment:
logging AND returning an error is a bad practice (the error most likely will be logged outside, which will cause log duplication).

Reply:
Yup, I'll improve the error approach in the follow-up PRs.
Right now I did it this way, since we log all the errors for failed constructors, but here it's slightly different.

Reply:
I was able to get rid of these lines completely.